scapy/contrib/isotp.py

```diff
@@ -57,289 +54,1346 @@ def answers(self, other):
+                    # A timeout happened
+                    callback()
+
+        def start(self, timeout):
+            """Starts the timer with the provided timeout, in seconds."""
+            if not self._completed:
+                raise Scapy_Exception("Timer was already started")
+
+            self._event.clear()
+            self._thread = Thread(target=ISOTPSocketImplementation.Timer._wait,
+                                  args=(self, self._event, timeout, self._callback))
+            self._completed = False
+            self._thread.start()
+
+        def cancel(self):
+            """This method can be used to stop the timer without executing the
+            callback."""
+            self._event.set()
+            if self._thread is not None:
+                self._thread.join()
+
+    class BlockingCallback:
+        """
+        Utility class to create callback objects that can be waited on until
+        they are executed on another thread.
+        """
+
+        def __init__(self):
+            self._event = Event()
+            self.args = None
+            self.kwargs = None
+
+        def wait(self, timeout=None):
+            """Wait until this object is called."""
+            return self._event.wait(timeout)
+
+        def callback(self, *args, **kwargs):
+            self._event.set()
+            self.args = args
+            self.kwargs = kwargs
+
+    def __init__(self,
+                 sendfunc,
+                 extended_addr=None,
+                 extended_rx_addr=None,
+                 rx_block_size=0,
+                 rx_separation_time_min=0):
+        """
+
+        :param sendfunc: Function that will be called whenever this object
+                decides that a CAN frame should be sent.
+        :param extended_addr: Extended Address byte to be added at the
+                beginning of every CAN frame _sent_ by this object. Can be None
+                in order to disable extended addressing on sent frames.
+        :param extended_rx_addr: Extended Address byte expected to be found at
+                the beginning of every CAN frame _received_ by this object. Can
+                be None in order to disable extended addressing on received
+                frames.
+        :param rx_block_size: Block Size byte to be included in every Control
+                Flow Frame sent by this object. The default value of 0 means
+                that all the data will be received in a single block.
+        :param rx_separation_time_min: Time Minimum Separation byte to be
+                included in every Control Flow Frame sent by this object. The
+                default value of 0 indicates that the peer will not wait any
+                time between sending frames.
+        """
+
+        self.sendfunc = sendfunc
+
+        self.extended_rx_addr = extended_rx_addr
+        self.ea_hdr = b""
+        if extended_addr is not None:
+            self.ea_hdr = struct.pack("B", extended_addr)
```

Comment:
Personal preference: use `chb(x)` rather than `struct.pack("B", x)`.
`chb` can be imported from `scapy.compat`

Reply:
FWIW, `chb()` seems much slower:
```Python
from timeit import timeit
timeit("chb(123)", setup="from scapy.compat import chb")
# 0.6313017439097166
timeit("struct.pack('B', 123)[0]", setup="import struct")
# 0.17192010302096605
```

Reply:
Nice catch, please ignore my comment